prov/gni/include/gnix_buddy_allocator.h

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright (c) 2015 Los Alamos National Security, LLC. All rights reserved.
+ * Copyright (c) 2015 Cray Inc.  All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef _GNIX_BUDDY_ALLOCATOR_H_
+#define _GNIX_BUDDY_ALLOCATOR_H_
+
+#include "fi_list.h"
+#include "gnix_bitmap.h"
+#include "gnix_util.h"
+#include <math.h>
+#include <stdlib.h>
+
+#define MIN_BLOCK_SIZE 16
+
+/* evaluates to zero if X is not a power of two, otherwise evaluates to X - 1 */
+#define IS_NOT_POW_TWO(X) (((X) & (~(X) + 1)) ^ (X))
```

Comment:
Could you elaborate?

> On Jan 7, 2016, at 5:34 PM, James Swaro <notifications@github.com> wrote:
> 
> In prov/gni/include/gnix_buddy_allocator.h:
> 
> > + * SOFTWARE.
> > + */
> > +
> > +#ifndef _GNIX_BUDDY_ALLOCATOR_H_
> > +#define _GNIX_BUDDY_ALLOCATOR_H_
> > +
> > +#include "fi_list.h"
> > +#include "gnix_bitmap.h"
> > +#include "gnix_util.h"
> > +#include <math.h>
> > +#include <stdlib.h>
> > +
> > +#define MIN_BLOCK_SIZE 16
> > +
> > +/* evaluates to zero if X is not a power of two, otherwise evaluates to X - 1 */
> > +#define IS_NOT_POW_TWO(X) (((X) & (~(X) + 1)) ^ (X))
> You might consider the following as well. It should result in fewer operations.
> 
> IS_POW_TWO(v) v && (v & (v - 1)))
> IS_NOT_POW_TWO(v) !IS_POW_TWO(v)
> 
> —
> Reply to this email directly or view it on GitHub.
>